### PR TITLE
Do not clear all results when searching from users address books

### PR DESCRIPTION
**Pull request by juliusknorr**
Steps to reproduce:
- Have any additional addressbook enabled (e.g. contacts app with a contact "A contact user" that has the email "contact@example.com")
- Restrict autocompletion in sharing
- Search for "A contact" in the sharing dialog

Before:
- The users own contact does not show up

After:
- The users own contact shows up while results from the system address book are still bocked

Problem: The wide matches in
https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/Collaboration/Collaborators/MailPlugin.php#L239-L240 are always cleared even though this should only happen for the system address book (actual users of the instance)

### CODE REVIEW

**Review by PVince81 — APPROVED** [review]
:+1:

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Can we have an integration test in https://github.com/nextcloud/server/blob/master/build/integration/collaboration_features/autocomplete.feature

**Comment by juliushaertl** [comment]
Yes, will look into the integration tests and the failing unit tests:

```

1) Test\Collaboration\Collaborators\MailPluginTest::testSearch with data set #6 ('test', array(array('uid3', 'User3 @ Localhost'), array('uid2', 'User2 @ Localhost', array()), array('uid1', 'User @ Localhost', array('username@localhost'))), false, array(array(), array(array())), false, false)
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'emails' => Array (
+        0 => Array (...)
     )
     'exact' => Array (...)
 )

/drone/src/tests/lib/Collaboration/Collaborators/MailPluginTest.php:133

2) Test\Collaboration\Collaborators\MailPluginTest::testSearch with data set #8 ('test@remote.com', array(array('uid3', 'User3 @ Localhost'), array('uid2', 'User2 @ Localhost', array()), array('uid1', 'User @ Localhost', array('username@localhost'))), false, array(array(), array(array(array('test@remote.com', 'test@remote.com', array(4, 'test@remote.com'))))), false, false)
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'emails' => Array (
+        0 => Array (...)
     )
     'exact' => Array (...)
 )

/drone/src/tests/lib/Collaboration/Collaborators/MailPluginTest.php:133

3) Test\Collaboration\Collaborators\MailPluginTest::testSearch with data set #12 ('user space@remote.com', array(array('uid3', 'User3 @ Localhost'), array('uid2', 'User2 @ Localhost', array()), array('uid1', 'User @ Localhost', array('username@localhost'))), false, array(array(), array(array())), false, false)
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'emails' => Array (
+        0 => Array (...)
     )
     'exact' => Array (...)
 )

/drone/src/tests/lib/Collaboration/Collaborators/MailPluginTest.php:133

--
```

**Comment by juliushaertl** [comment]
Unit tests adjusted and added intergration tests to cover the MailPlugin with and without enumeration

**Comment by juliushaertl** [comment]
@nickvergessen I'm a bit confused about the `shareapi_restrict_user_enumeration_full_match` when he parent sharing enumeration is enabled:

image

It seems to have no effect in the current UserPlugin/MailPlugin implementation. Is that only relevant if general sharee enumeration is disabled? Also the naming in the UI seems to be confusing as it mentions **allow**, but in the setting itself it is **restrict**.

**Comment by juliushaertl** [comment]
Basically with the above settings I'd expect to only get full matches, but it seems to return also wide matches for local users.

**Comment by nickvergessen** [comment]
Explained it a bit in the chat.
The main idea is that when enumeration is restricted, we need an option to say if you are allowed to share/autocomplete/etc when you match the full user id/display name, or if you have to have a common group/phone number match in order to be able to do so.

**Review by skjnldsv — APPROVED** [review]

File: lib/public/Files/Cache/ICache.php [hunk 1]
```diff
@@ -111,7 +111,7 @@ public function insert($file, array $data);
 	 * update the metadata of an existing file or folder in the cache
 	 *
 	 * @param int $id the fileid of the existing file or folder
-	 * @param array $data [$key => $value] the metadata to update, only the fields provided in the array will be updated, non-provided values will remain unchanged
+	 * @param array $data [$key => $value] the metadata to update, only thcccce fields provided in the array will be updated, non-provided values will remain unchanged
```
Comment (nickvergessen) [inline lib/public/Files/Cache/ICache.php:114]: ????

Comment (juliushaertl) [reply lib/public/Files/Cache/ICache.php]: Whoops, sorry not sure how that got in. Addressed.

**Comment by juliushaertl** [comment]
All tests pass now and remaining one seems unrelated when creating a folder in the sharing acceptance test.

**Pull request merged**